### PR TITLE
Cookbook normalize vector

### PR DESCRIPTION
**Pull request by ggsmith842**
I added a recipe for the Normalize a Vector task in the Vector and Matrix Operations section. It uses the standard library and I provided an example of the usage.

Functions included:

- `sum` - to sum a list of squares.
- `magnitude` - calls the function `Float.sqrt` on the summed squares list to return the magnitude. 
- `normalize` - calculates the unit vector by dividing each element of the vector, v by the magnitude.

### CODE REVIEW

File: data/cookbook/normalize-vector/00-stdlib.ml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+---
+packages: []
+---
+
+(**[sum] returns the sum of all elements in the list [lst] *)
```
Comment (cuihtlauac) [inline data/cookbook/normalize-vector/00-stdlib.ml]:
```suggestion
(* `sum` returns the sum of all elements in the list `lst` *)
```

File: data/cookbook/normalize-vector/00-stdlib.ml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+let sum lst = List.fold_left ( +. ) 0. lst
+
+(** returns the [magnitude] for a sum of squares [sqs]*)
```
Comment (cuihtlauac) [inline data/cookbook/normalize-vector/00-stdlib.ml]:
```suggestion
(* returns the `magnitude` for a sum of squares `sqs`*)
```

File: data/cookbook/normalize-vector/00-stdlib.ml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+let magnitude sqs = Float.sqrt (sum sqs)
+
+(** [normalize] accepts a list [v] of float values and returns the unit vector as a list*)
```
Comment (cuihtlauac) [inline data/cookbook/normalize-vector/00-stdlib.ml]:
```suggestion
(** `normalize` accepts a list `v` of float values and returns the unit vector as a list*)
```

File: data/cookbook/normalize-vector/00-stdlib.ml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+;;
+
+
+(* Example Usage *)
+let vector = [4.;6.;-1.]
+let vhat = normalize vector;;
+List.iter (Printf.printf "%.2f ") vhat
+
+(* Note the values are rounded to the 2nd decimal place for readability.
+          In practice, you should not round until your final calculation.*)
```
Comment (cuihtlauac) [inline data/cookbook/normalize-vector/00-stdlib.ml]:
```suggestion
(* Note the values are rounded to the 2nd decimal place for readability.
   In practice, you should not round until your final calculation.*)
List.iter (Printf.printf "%.2f ") vhat
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]
Thnaks @ggsmith842, this looks good to me. Unfortunately cookbook recipes don't support ocamlformat/odoc.

**Comment by ggsmith842** [comment]
Thanks @cuihtlauac! I added your suggestions removing the odoc comments.

File: data/cookbook/normalize-vector/00-stdlib.ml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+packages: []
+---
+
+(* `sum` returns the sum of all elements in the list `lst` *)
+let sum lst = List.fold_left ( +. ) 0. lst
+
+(* returns the `magnitude` for a sum of squares `sqs`*)
+let magnitude sqs = Float.sqrt (sum sqs)
+
+(** `normalize` accepts a list `v` of float values and returns the unit vector as a list*)
```
Comment (cuihtlauac) [inline data/cookbook/normalize-vector/00-stdlib.ml]:
```suggestion
(* `normalize` accepts a list `v` of float values and returns the unit vector as a list*)
```

With two stars, you get an markdown item, sorry about that

File: data/cookbook/normalize-vector/00-stdlib.ml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+let normalize v =
+  let sqs = List.map (fun x -> x *. x) v in
+  let m = magnitude sqs in
+  List.map (fun x -> x /. m) v
+;;
```
Comment (cuihtlauac) [inline data/cookbook/normalize-vector/00-stdlib.ml]:
```suggestion
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

File: data/cookbook/normalize-vector/00-stdlib.ml [hunk 7]
```diff
@@ -0,0 +1,26 @@
+---
+packages: []
+---
+
+(* `sum` returns the sum of all elements in the list `lst` *)
+let sum lst = List.fold_left ( +. ) 0. lst
+
+(* returns the `magnitude` for a sum of squares `sqs`*)
```
Comment (cuihtlauac) [inline data/cookbook/normalize-vector/00-stdlib.ml]:
```suggestion
(* Returns the `magnitude` for a sum of squares `sqs`*)
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

**Review by cuihtlauac — APPROVED** [review]

**Review by cuihtlauac — APPROVED** [review]
This looks good to me, thanks @ggsmith842

**Review by christinerose — APPROVED** [review]

**Review by sabine — APPROVED** [review]
Thanks @ggsmith842! LGTM.

Merging after line editing confirmed by @christinerose.

**Review by christinerose — APPROVED** [review]

**Comment by RCHG** [comment]
Thanks for this cookbook. Two minor comments:

(1) In the current form probably would fail,
```ocaml
let () =
  let vector = [4.;6.;-1.]
  let vhat = normalize vector;;
  List.iter (Printf.printf "%.2f ") vhat
```
so maybe it is better like:
```ocaml
let () =
  let vector = [4.;6.;-1.] in 
  let vhat = normalize vector in 
  List.iter (Printf.printf "%.2f ") vhat
```
(2) Sorry to be a bit picky but formally speaking this is using the definition of the Euclidean norm or  2-norm. The general definition of normalization is named p-norm which is implemented in the Owl library: `Owl.Owl_linalg_generic.vecnorm ~p x` that has an optional argument p.

Because the Euclidean norm is much more used that others, then it is totally fine to just it in a cookbook. Still, I would indicate that fact in the comments describing the code. So instead of:
```ocaml
(* normalize accepts a list v of float values and returns the corresponding unit vector as a list. *)
```
we could say
```ocaml
(* normalize accepts a list v of float values and returns the corresponding unit vector as a list (using the Euclidean norm as the norm of the vector v) *)
```

**Pull request merged**